core/src/main/java/hudson/FilePath.java

```diff
@@ -1166,13 +1175,60 @@ public Boolean invoke(File f, VirtualChannel channel) throws IOException, Interr
      * Deletes this directory, including all its contents recursively.
      */
     public void deleteRecursive() throws IOException, InterruptedException {
-        act(new SecureFileCallable<Void>() {
-            private static final long serialVersionUID = 1L;
-            public Void invoke(File f, VirtualChannel channel) throws IOException {
-                deleteRecursive(deleting(f));
-                return null;
+        new DisposeDirectory(this).perform();
+    }
+
+    private static final class DisposeDirectory extends MasterToSlaveCallable<Boolean, IOException> implements Disposable {
+        private static final long serialVersionUID = 1L;
+
+        private final FilePath file;
+
+        private DisposeDirectory(FilePath file) {
+            this.file = file;
+        }
+
+        public boolean dispose() throws Exception {
+            Boolean exists = file.act(this);
+            assert exists != null;
+            return !exists;
+        }
+
+        public String getDisplayName() {
+            return "Directory " + file + " over " + file.channel;
+        }
+
+        private void perform() throws IOException, InterruptedException {
+            FilePath tempFile = file.getParent().child(file.getBaseName() + "_async_cleanup_" + System.currentTimeMillis());
+            boolean renamed = false;
+            try {
+                file.renameTo(tempFile);
+                renamed = true;
+
+                Register register = new Register(this);
+                if (Jenkins.getInstance() != null) {
+                    register.call(); // Register locally
+                    return;
+                }
+
+                // Master on the other end confirmed registration
+                if (file.channel != null && file.channel.call(register)) return;
+            } catch (IOException ex) {
+                // Failed to rename or schedule remotely
             }
-        });
+
+            // Fallback to synchronous delete when not on master or not talking to master
+            if (renamed) {
+                tempFile.renameTo(file);
+            }
+
+            file.act(this);
```

Comment:
so if the rename worked on windows you can delete the directory recursivly - so this just adds unnessecary overhead.
If you can't rename the directory on windows your not going to be able to delete it either.